### ISSUE TITLE
Not picking up node.exe despite being in path

### ISSUE THREAD

**Issue by WhatFreshHellIsThis** [tjanczuk/iisnode#302]
## Windows server 2012 64bit

This is an error I receive:
## The iisnode module is unable to start the node.exe process. Make sure the node.exe executable is available at the location specified in the system.webServer/iisnode/@nodeProcessCommandLine element of web.config. By default node.exe is expected in one of the directories listed in the PATH environment variable.

This is wrong, node.exe _is_ in my path and I can open a command prompt and run node.exe no problems, so I went into the web.config area indicated in the error message and it states under nodeProcessCommandLine: node.exe which should work but doesn't.   If I change it to the full path of the node executable (C:\Program Files\nodejs\node.exe) it suddenly works.

**Comment by calebaker** [comment]
I'm getting exact same error on Server 2008 R2, I've tried x86 and x64 versions of node.exe.

**Comment by merkt** [comment]
Modifying my PATH variable to "%ProgramFiles%\nodejs\" (as opposed to "C:\Program Files\nodejs\node.exe") and restarting WAS as @rramachand21 suggested seemed to resolve this for me. This was with the x64 version of Node on Server 2008 R2.

**Comment by fraguada** [comment]
Just to chime in since I had a similar issue.  I have a webapp on Azure that was working fine until I went to update some dlls to x64.  Seems they could not be loaded.  As others have stated, Azure runs x86 node.js no matter the platform you specify.  One can upload their desired version of node.js and use nodeProcessCommandLine="nodeDir" to point iisnode there.  My issue was that iisnode seemed to automatically generate an iisnode.yml which I did not notice.  This file was still pointing to the x86 version of node.js and even though the web.config was correctly configured to point to my custom path, iisnode.yml was overriding that.  The fix was to remove the nodeProcessCommandLine attribute in web.config, and have iisnode.yml point to the right directory.  I hope that helps someone!

**Comment by amcdnl** [comment]
I tried @merkt example but failed, ended up having to manually put `<iisnode nodeProcessCommandLine="&quot;%programfiles%\nodejs\node.exe&quot;" />` in the webconfig. ( x64 node on Win10 )

**Comment by sylvain-hamel** [comment]
I found out that if you use `nvm` to install node, you end up with a symlink to a folder somewhere in `C:\Users\YOU\AppData\Roaming\nvm\v6.9.1` and the account running the AppPool does not have access to this location.

**Comment by devotis** [comment]
I also have node on my path:

C:\Windows\system32>path
PATH=[other paths that I hid for this post];C:\Python27;C:\node\nvm;C:\node\nodejs
C:\Windows\system32>node -v
v6.9.2

And still received this error when running a website:

_The iisnode module is unable to start the node.exe process. Make sure the node.exe executable is available at the location specified in the system.webServer/iisnode/@nodeProcessCommandLine element of web.config. By default node.exe is expected in one of the directories listed in the PATH environment variable._

Restarting WAS (net stop was /y & net start w3svc) didn't make a difference.

I had to add nodeProcessCommandLine="C:\node\nodejs\node.exe" to web.config for the website to work.

However after a reboot. I found that I could remove the nodeProcessCommandLine="nodedir" part and after `iisreset`, I tried the website again and no error.

TL;DR
Reboot the server and you don't need to add nodeProcessCommandLine="nodeDir" as long as node is on your path. Test that with running `node -v` on the command line. And also I notices that right after the reboot. The websites are malfunctioning in general. Just wait for it a couple of minutes and try again. Or you might wrongfully come to the conclusion there's something wrong with your configuration. This is Windows Server 2008 R2 and IIS 7.5.

**Comment by dmidz** [comment]
In fact I got exactly the sames problems : paths in env PATH or nodeProcessCommandLine seemed wrong and windows shortcuts or symlinks not working. As suggested sylvain-hamel, it is all a permission problem, everything works correctly as soon as :
- you install nvm in a directory with no spaces such program files because of nvm switching not working, nor in your user dir because of below permission, so install it at the root of C: for ex.
- you give read & execution permissions to this nvm install directory to the IIS user "IIS_IUSRS"
![nvm_install_permissions](https://user-images.githubusercontent.com/1290511/27032540-995a9618-4f75-11e7-81fa-53fd609bc860.png)

**Comment by Foxhound013** [comment]
I'm assuming everyone in this thread has figured out the issue? I've read through everything and checked on everything I could possibly check. My web config file looks like the following.

`<configuration>
  <system.webServer>
    
    <handlers>
      <add name="iisnode" path="app.js" verb="*" modules="iisnode" />
    </handlers>

    <rewrite>
      <rules>
        <rule name="sendToNode">
          <match url="/*" />
          <action type="Rewrite" url="routes/road" />
        </rule>

        <rule name="DynamicContent">
          <conditions>
            <add input="{REQUEST_FILENAME}" matchType="IsFile" negate="True"/>
          </conditions>
          <match url="/*" />
          <action type="Rewrite" url="app.js"/>
        </rule>
        
      </rules>
    </rewrite>

    <security>
      <requestFiltering>
        <hiddenSegments>
          <add segment="node_modules" />
        </hiddenSegments>
      </requestFiltering>
    </security>

		<iisnode nodeProcessCommandLine="%C:\Program Files\nodejs\node.exe%;" nodeProcessCountPerApplication="8" watchedFiles="*.js;node_modules\*;routes\*.js;views\*.hbs"/> 
	</system.webServer> 
</configuration>`

If I switch this line to <action type="Rewrite" url="routes/road.js" /> I get the content of my road.js file served to me. Without it I get told that node.js isn't able to start. I've checked permissions of nvm and that seems to be fine, it has full permission.

I installed nvm under the base C directory (C:/nvm) and my path looks like the following

%NVM_SYMLINK%;%node.exe%;

Any help would be greatly appreciated.

**Comment by dmidz** [comment]
@Foxhound013
You sould have also env variables NVM_HOME (C:\nvm) & NVM_SYMLINK (C:\Program Files\nodejs) created with the install. My PATH variable contains also a ref to NVM_HOME. Your %node.exe% seems wrong, enclosed name in % is a ref to an env variable, I bet you have no variable named "node.exe", I do not have it in my PATH :
`...;%NVM_HOME%;%NVM_SYMLINK%;...`

I get problems when watchedFiles contained inexistants dirs also...
I see also that your match condition seems useless and I do not use nodeProcessCommandLine parameter...it will use env variables...
```
<configuration>
    <system.webServer>
        <handlers>
            <add name="iisnode" path="app.js" verb="*" modules="iisnode" />
        </handlers>
	<rewrite>
		<rules>
               		...
               		<rule name="DynamicContent">
	                	<conditions>
		                        <add input="{REQUEST_FILENAME}" matchType="IsFile" negate="True" />
              			</conditions>
              			<action type="Rewrite" url="app.js" />
	                </rule>
		</rules>
	</rewrite>
	<iisnode watchedFiles="web.config;*;app.js"
		 node_env="%node_env%"
		 nodeProcessCountPerApplication="1"
		 maxConcurrentRequestsPerProcess="1024"
		 maxNamedPipeConnectionRetry="100"
		 namedPipeConnectionRetryDelay="250"
		 maxNamedPipeConnectionPoolSize="512"
		 maxNamedPipePooledConnectionAge="30000"
		 asyncCompletionThreadCount="0"
		 initialRequestBufferSize="4096"
		 maxRequestBufferSize="65536"
		 uncFileChangesPollingInterval="5000"
		 gracefulShutdownTimeout="60000"
		 loggingEnabled="true"
		 logDirectory="logs"
		 debuggingEnabled="false"
		 debugHeaderEnabled="false"
		 debuggerPortRange="5058-6058"
		 debuggerPathSegment="debug"
		 maxLogFileSizeInKB="128"
		 maxTotalLogFileSizeInKB="1024"
		 maxLogFiles="20"
		 devErrorsEnabled="true"
		 flushResponse="false"
		 enableXFF="true"
		 promoteServerVars=""
		 configOverrides="iisnode.yml" />
	...
```

**Comment by abduljamalbasha** [comment]
above config file working fine for me thanx.

**Comment by Foxhound013** [comment]
@dmidz ,

I checked what you had mentioned about my path settings. This is what I've got, I've also removed the %node.exe%.

- I have NVM_HOME at C:\nvm
- I have NVM_SYMLINK at C:\Program Files\nodejs
- My path variable is now as follows . . . %NVM_HOME%;%NVM_SYMLINK%;

I've made some changes to my web.config file based on your suggestions and suggestions I found in a few other places. Here is my new web.config file. 

`<?xml version="1.0" encoding="utf-8"?>
<configuration>
  <system.webServer>

    
    <security>
      <requestFiltering>
        <hiddenSegments>
          <remove segment="bin" />
        </hiddenSegments>
      </requestFiltering>
    </security>

    <handlers>
      
      <add name="iisnode" path="/bin/www" verb="*" modules="iisnode"/>
    </handlers>
    <rewrite>
      <rules>
        <rule name="NodeInspector" patternSyntax="ECMAScript" stopProcessing="true">
          <match url="^bin\/www\/debug[\/]?" />
        </rule>

        
        <rule name="StaticContent">
          <action type="Rewrite" url="public{REQUEST_URI}"/>
        </rule>

        
        <rule name="DynamicContent">
          <conditions>
            <add input="{REQUEST_FILENAME}" matchType="IsFile" negate="True"/>
          </conditions>
          <action type="Rewrite" url="/bin/www"/>
        </rule>
      </rules>
    </rewrite>

    <iisnode watchedFiles="web.config;iisnode.yml;*.js;node_modules\*;routes\*.js;views\*.hbs"
		 recycleSignalEnabled="true"
	 	logDirectory="logs"
	 	nodeProcessCommandLine="%programfiles%\nodejs\node.exe"
		 nodeProcessCountPerApplication="1"
		 maxConcurrentRequestsPerProcess="1024"
		 maxNamedPipeConnectionRetry="100"
		 namedPipeConnectionRetryDelay="250"
		 maxNamedPipeConnectionPoolSize="512"
		 maxNamedPipePooledConnectionAge="30000"
		 asyncCompletionThreadCount="0"
		 initialRequestBufferSize="4096"
		 maxRequestBufferSize="65536"
		 uncFileChangesPollingInterval="5000"
		 gracefulShutdownTimeout="60000"
		 loggingEnabled="true"
		 debuggingEnabled="false"
		 debugHeaderEnabled="false"
		 debuggerPortRange="5058-6058"
		 debuggerPathSegment="debug"
		 maxLogFileSizeInKB="128"
		 maxTotalLogFileSizeInKB="1024"
		 maxLogFiles="20"
		 devErrorsEnabled="false"
		 flushResponse="false"
		 enableXFF="false"
		 promoteServerVars="SERVER_NAME" />
  </system.webServer>
</configuration>
`

I am now having issues with getting a **404.8 error: The request filtering module is configured to deny a path in the URL that contains a hiddenSegment section.**

Here is how I've got my things set up in IIS, I right click default web site and select add application. From this point I chose the path to my applications files. I have the web.config in the root of the folder. From what I understand, there are issues with the application starting from within the bin folder but shouldn't removing the security on bin fix this?

My file structure is as follows ....

- bin
   -www
-node_modules
   -Various npm packages like express
-public
   -various scripts
-routes
   -routes I'm interested in
-views
   -my handlebars templates
.bowerrc
app.js
package-lock.json
package.json
web.config

Have I set up the directory incorrectly or is the issue still in my web.config?

**Comment by dmidz** [comment]
@Foxhound013
before customizing so much your working way, you might go step by step following basis exactly until it works, then adding your preferences one by one :
- try comment your security filter for the moment
- caution with rules, for the DynamicContent to be caught, previous ones must match/unmatch correctly, check my complete rules part, it tries to catch static file if found in public dir, and stop rules in this case (stopProcessing), and catch last one DynamicContent only if no previous caught (I wiped out debug part for the moment) :

```
...
		<rules>
    			<clear/><!-- ////// IMPORTANT : it allows to clear inherited things that would break things for rewrite rules //////-->
			<!-- {DOCUMENT_ROOT} = root dir of website, {APPL_PHYSICAL_PATH} = root dir of sub app -->
			<!-- __ check if could catch a real file in public dir -->
			<rule name="StaticContent" stopProcessing="true">
    				<match url="^.+"/>
				<conditions logicalGrouping="MatchAll">
		                        <add input="{APPL_PHYSICAL_PATH}public/{R:0}" matchType="IsFile" />
				</conditions>
				<action type="Rewrite" url="public/{R:0}" />
			</rule>
			<!-- All other URLs are mapped to the Node.js application entry point -->
			<rule name="DynamicContent">
				<action type="Rewrite" url="app.js" />
				<serverVariables>
					<!-- # WARN : outboundRules is incompatible with Gzip so must remove header Accept-Encoding below -->
					<set name="HTTP_ACCEPT_ENCODING" value="" />
				</serverVariables>
			</rule>
		</rules>
```
- what does contain your www file ? it mounts app.js ?
- your IIS website or application document root should point to your App dir, considering the paths you set. The path you mention is the default IIS document path, yours seems not been configured, use IIS manager
- I forgot to mention : as we set a logDirectory parameter, the directory must exists, create a "logs" dir on the same level as public, this one will have to be filtered if any...but normally the rules I gave you prevent viewing anything other than public files, or be routed to your nodejs script...
logs will contains any stdout or stderr (console.*) coming from your nodejs script. If the directory is still empty, it means nodejs script has not been ran yet.

**Comment by Foxhound013** [comment]
@dmidz ,

- My www file is the entry point for the application (I believe you're right, it mounts app.js). I had used express-generator to generate my file structure at the beginning of this project. 
- I had set the correct path to the apps directory when I created the application in IIS but it seems to be disregarding it. If I change bin/www to app.js as follows,
`<rule name="DynamicContent">
          <conditions>
            <add input="{REQUEST_FILENAME}" matchType="IsFile" negate="True"/>
          </conditions>
          <action type="Rewrite" url="**app.js**"/>
        </rule>`
then IIS serves up my source code from app.js but doesn't route me anywhere.

- I did at one point have error messages coming to a folder called iisnode inside of my root directory, I'll see if I can get back to that point.

**Comment by jaysingit** [comment]
By setting the values <iisnode nodeProcessCommandLine="C:\Program Files\nodejs\node.exe" />   in C:\Program Files\iisnode\www\configuratio\web.config is not always fixing the issue . Please follow the below steps and try to set in IIS server level.

Step 1: Select IIS server and open Configuration editor https://www.screencast.com/t/SWcy4C5m 
Step 2: Select the section system.webServer/iisnode and set the nodeProcessCommandLine value to            node.exe full path. https://www.screencast.com/t/e8N6MFeWEueS 

This fixed my issue.
